src/xview/moxftp-2.2/Wcl-2.4/Xmp/Xmp.man
.COMMENT SCCS_data: @(#) Xmp.man 1.3 92/11/02 08:36:37
.TH "Xmp" 1 "1 March 1992"
.SH NAME
Xmp \- Motif Public Library
.SH SYNOPSIS
.ta 1.5i 2.5i
.nf
.ft B
#include <X11/Intrinsic.h>
#include <X11/Wc/WcCreate.h>
#include <X11/Xmp/Xmp.h>

/* All Wcl applications should provide at least the Wcl options:
*/
static XrmOptionDescRec options[] = {
    ... application specific command line options ...
    WCL_XRM_OPTIONS
};

main ( argc, argv )
    int    argc;
    char*  argv[];
{
    /*  -- Intialize Toolkit creating the application shell
    */
    Widget appShell = XtInitialize (
        argv[0], WcAppClass( argc, argv ),      /* app name & class  */
        options, XtNumber(options),             /* argv option descr */
        &argc, argv
    );
    XtAppContext app = XtWidgetToApplicationContext(appShell);

    /*  -- Register application specific callbacks and widget classes
    */
    RegisterApplication ( app );

    /*  -- Register all widget classes and constructors
    */
    XmpRegisterMotif( app );

    /*  -- Create widget tree below toplevel shell using Xrm database
    */
    if ( WcWidgetCreation ( appShell ) )
        exit(1);

    /*  -- Realize the widget tree
    */
    XtRealizeWidget ( appShell );

    /*  -- Catch `Close' from mwm menu on top level shell.
    */
    XmpAddMwmCloseCallback( appShell, DeleteWindowCB, clientData );

    /*  -- and finally, enter the main application loop
    */
    XtMainLoop ( );
}
.fi
.ft R

.SH LINK BY
.RS
.nf
.ft B
example% ld -o prog prog.o -lXmp -lXm -lWc <dynLinkLib> -lXmu -lXt -lX11
.ft R
.RE
.LP 
On SunOS or SVR4 machines, the <dynLinkLib> is named -ldl.
Other architectures cannot use the dynamic linking capabilites of
Wcl, and therefore do not need any dynamic link library.
.SH DESCRIPTION
.LP
.I Xmp
is intended to be used with
.I Wcl
and the Motif widget set.  Xmp provides a procedure for registering all
Motif widgets, and convenience functions frequently needed by Motif
applications.
.SH REGISTRATION PROCEDURES
.LP
.I Wcl
must be told the names of all widgets and widget constructors before
it can create widgets as specified in the resource file of client
applications.
.I Xmp
provides a registration procedure for registering all Motif widgets.
For backward compatibility with previous releases of Wcl, this
registration procedure has three names: MriRegisterMotif(), 
XmpRegisterAll(), and XmpRegisterMotif().  These procedures
are exactly equivalent.
.TP
.B void XmpRegisterMotif( XtAppContext )
This procedure registers all of the Motif widget class pointer names,
widget class names, and widget constructor function names with Wcl.
It also registers all Xmp converters and action procedures with Xt,
and all Xmp callback procedures with Wcl.
.SH CONVERTERS
.LP
.I Xmp
provides converters which are useful for specifying look and feel
attributes of Motif user interfaces from resource files.  It causes
the WcCvtStringToWidget() converter to be invoked for setting
constraints and menuIds.  WcCvtStringToWidget() uses WcFullNameToWidget()
to resolve the name, so the Wcl relative path naming conventions can
be used to navigate the widget tree to find a specific widget.
.LP
.I Xmp
causes the XmpCvtStringToXmString()
converter to be invoked to provide multi-line labels to Motif
widgets.  This converter uses XmStringCreateLtoR() and the
XmSTRING_DEFAULT_CHARSET to convert strings in the resource database
into XmStrings.  XmStringCreateLtoR() recognizes the two characters
"\n" (backslash followed by a lower case "n"), and causes the XmString
to break to a new line.  Hence, "Hello\nThere" ends up being a two
line label on an XmLabel.
.SH ACTIONS AND CALLBACKS
.LP
Most Xmp actions are available as callbacks, and all callbacks are
available as actions.  The provided procedures, as named in the
resource database, are as follows:
.nf
.ft B
XmpFixTranslations( widget [widget ...] )
XmpAddMwmCloseCallback( widget callback(args) [callback(args) ...] )
XmpAddTabGroup( [widget] )
XmpTableChildConfig( widget col row [h_span [v_span [opts]]] )
XmpTableChildPosition( widget col row )
XmpTableChildResize( widget h_span v_span )
XmpTableChildOptions( widget options )
XmpPopup( widget )
.ft R
.fi
.LP
Each callback procedure can also be named in resource files as
the name of the procedure followed by CB: i.e, XmpFixTranslations()
can also be invoked as a callback by giving the name XmpFixTranslationsCB()
for a callback resource value.  Also, each action procedure can also
be named by appending ACT to the procedure name.  Normally, I just use the
procedure name, so it is less trouble to change behavior from callbacks
to actions triggered by translations or accelerators, and vice versa.
The CB and ACT endings are maintained for backward compatibility, and
because some people like to be more explicit.  It is all a matter of taste.
.LP
Arguments to the procedures are provided as strings, the characters
between the parenthesis following the procedure name.  If no parenthesis
follow the procedure name, then a NULL string is passed to the procedure.
Some procedures require arguments, some provide reasonable defaults when
no arguments are given.  For example:
.RS
.nf
.ft B
*foo.wcCreate:         XmText
*foo.wcCallback:       XmpFixTranslations( this ^that )

*bar.wcCreate:         XmPushButton
*bar.activateCallback: XmpAddTabGroup
.ft R
.RE
.LP
Each
.I Xmp
procedure is discussed in detail below.
.TP
.B XmpFixTranslations( widget [widget ...] )
This works on XmText widgets only, and has the effect of fixing
translations which get hammered when XmText widgets are put into
dialogs.  Normally, pressing newline in a multi-line text widget
does a newline action.  In a dialog, pressing return causes the
dialog to go away.  
.TP
.B XmpAddMwmCloseCallback( widget callback(args) [callback(args) ...] )
The callbacks are invoked in an Xt dependent order when the named widget
receives the WM_DELETE_WINDOW or WM_SAVE_YOURSELF protocol event.  Each
of the named callbacks are added to the widget using XmAddProtocolCallback(). 
.TP
.B XmpAddTabGroup( [widget] )
The named widget, or the widget providing the context for the callback if
no widget is named, is passed to XmAddTabGroup().  The effect is simple:
if the widget is a manager widget, then all the children are put in a
tab group together.  If the widget is a primitive, then that widget
is in its own tab group.
.TP
.B XmpPopup( widget )
This is only available as an action triggered due to translations of
accelerators, it cannot be invoked as a callback.  The argument is the
name of a menu pane (usually, a widget which was created with
XmCreatePopupMenu()).  The menu is positioned using XmMenuPosition()
which is the same procedure used by Motif itself.  This depends on
the existance of an XmRowColumn somewhere in the interface.  If
you have a menuBar, then this will work.  This is because XmRowColumn
provides all sorts of wierd and wonderful event handlers for dealing
with menus.
.TP
.B XmpTableChildConfig( w col row [h_span [v_span [opts]]] )
This allows a child of an XmpTable to be moved to a new row or
column, to be given a different horizontal or vertical span, and
to change the justification and re-size options of the child.
.TP
.B XmpTableChildPosition( w col row )
This allows a child of an XmpTable to be moved to a new row or 
column.
.TP
.B XmpTableChildResize( w h_span v_span )
This allows a child of an XmpTable to be given a different horizontal
or vertical span.
.TP
.B XmpTableChildOptions( w options )
This allows a child of an XmpTable to be given new
justification and re-size options.
.SH "SEE ALSO"
.BI XmpTable (3),
.BI Mri (1),	
.BI Wcl (3),
.BI X (1)
.SH BUGS
.LP
realizeCallback, destroyCallback, popupCallback, and popdownCallback
MUST NOT be specified using resource values for shell widgets derived
from Motif Vendor Shell in most releases of Motif.  Sorry, there is no
way I can see for Wcl to work around this bug in the list of resources
fetched by the Motif Vendor Shell Extension VendorExtInitialize()
procedure.
.LP
If you want to specify these resources, specifiy a wcCallback resource
and add the callbacks using WcAddCallbacks().
.LP
Explanation for Xt and Motif gurus with source code: Xt has fetched
these resources for the shell, and compiled the XtCallbackRec array
generated by WcCvtStringToCallback into the
InternalCallbackRec/XtCallbackRec "thang."  This happens in
Xt'Create.c'_XtCreate in R5.  _XtCreate then allows the widget to
initialize itself, calling CallInitialize().  This eventually invokes
the Motif Vendor Shell initialization, which invokes the Motif Vendor
Shell Extension initialization.  Here, Motif incorrectly fetches again
the above callback resources, applying the resources straight into the
"new" widget with its already compiled callback lists.  This causes the
compiled callback list storage to be overwritten with the original,
uncompiled form previously generated by WcCvtStringToCallback.  The
next XtAddCallback or XtCallAllCallbacks or XtSetValues on these
resources will (or should!) fail due to mis-interpreting the callback
list (the first element will again be an XtCallbackRec instead of an
InternalCallbackRec).  The fix is messy:  somehow treat the callback
resources as uniquely as Xt does everywhere else: fetch into a
different (local) Widget instead of new, and then traverse the list of
callbacks, invoking XtAddCallback( new, <callbackName>, cb->callback,
cb->closure ), and then free up that locally allocated dummy widget.
Yuck!
.SH AUTHOR
David E. Smyth (David.Smyth@sniap.mchp.sni.de) at Siemens Nixdorf
Informationssysteme AG, Munich Germany.


src/NeWS
../NeWS

src/xview/moxftp-2.2/Wcl-2.4/Xmp/Xmp.c
#include <X11/Xmp/COPY>
/*
* SCCS_data: @(#) Xmp.c	1.5 92/10/28 07:50:36
*
* Xmp - Motif Public Utilities for Wcl - Xmp.c
*
* This file contains Motif specific converters, actions, and convenience
* functions for Motif and Xmp widgets.
*
*******************************************************************************
*/

#include <X11/IntrinsicP.h>

#ifdef sun
#include <X11/ObjectP.h>        /* why don't they just use X from mit!?! */
#include <X11/RectObjP.h>
#endif

#include <X11/StringDefs.h>
#include <Xm/Xm.h>
#include <Xm/RowColumnP.h>	/* RC_Type */
#include <Xm/Text.h>		/* xmTextWidgetClass */
#include <X11/Shell.h>		/* wmShellWidgetClass */
#include <X11/Wc/WcCreateP.h>

#include <X11/Xmp/Table.h>
#include <X11/Xmp/XmpP.h>

/******************************************************************************
* Private_macro_definitions.
******************************************************************************/

/* The `done' macro, used by the converters, is defined in WcCreateP.h */

/******************************************************************************
**  Motif specific resource converters
******************************************************************************/

/*
    -- Convert String To XmString
*******************************************************************************
    This conversion converts strings into XmStrings.  This one overrides
    the one provided by Motif which calls XmCreateString.  Instead, this
    one uses XmCreateStringLtoR which allows multi-line strings to be
    created by embedding "\n" in a resource specification.  In order to
    allow this to be used, XmpRegAll MUST install this converter AFTER
    Motif has installed its broken converter.
*/

void XmpCvtStringToXmString (args, num_args, fromVal, toVal)
    XrmValue *args;
    Cardinal *num_args;
    XrmValue *fromVal;
    XrmValue *toVal;
{
    XmString xmString;
    xmString = XmStringCreateLtoR( fromVal->addr, XmSTRING_DEFAULT_CHARSET );
    done( XmString, xmString );
}

void XmpFixTranslationsACT( w, unusedEvent, params, num_params )
    Widget      w;
    XEvent     *unusedEvent;
    String     *params;
    Cardinal   *num_params;
{
    static int            firstTime = 1;
    static XtTranslations transTable;
    static char           fixedTranslations[] = "<Key>Return: newline()";

    int     i;

    if (firstTime)
    {
	firstTime = 0;
	transTable = XtParseTranslationTable(fixedTranslations);
    }

    if ( *num_params == 0 )
    {
	WcWARN( w, "XmpFixTranslations", "usage",
		"Usage: XmpFixTranslations( widget [widgetName] ...)");
	return;
    }

    for ( i = 0 ; i < *num_params ; i++ )
    {
	Widget toFix = WcFullNameToWidget( w, params[i] );

        if (!toFix)
        {
	    WcWARN1( w, "XmpFixTranslations", "notFound",
		   "XmpFixTranslations(%s) - Widget not found", params[i] );
	    return;
        }

	if (XtIsSubclass( toFix, xmTextWidgetClass ))
	    XtOverrideTranslations( toFix, transTable );
	else
	{
	    WcWARN1( w, "XmpFixTranslations", "notText",
		   "XmpFixTranslations(%s) - Not an XmText widget", params[i] );
	    return;
        }
    }
}

void XmpFixTranslationsCB( w, widgetNames, ignore )
    Widget	w;
    XtPointer	widgetNames;
    XtPointer	ignore;
{
    WcInvokeAction( XmpFixTranslationsACT, w, widgetNames );
}


/*
    -- Action to popup a Motif PopupMenu
*******************************************************************************
*/

void XmpPopupACT( w, event, params, num_params )
    Widget      w;
    XEvent     *event;
    String     *params;
    Cardinal   *num_params;
{
    Widget      menuPane;

    if (*num_params < 1)
	return;

    if (menuPane = WcFullNameToWidget ( w, params[0] ))
    {
        XmMenuPosition( menuPane, event );
        XtManageChild( menuPane );
    }
}


/*
    -- Provide `Close' callback
*******************************************************************************
    Provide a callback which gets invoked when the user select `Close' from
    the Mwm frame menu on the top level shell.  MUST be done after shell widget
    is realized!
*/

void XmpAddMwmCloseCallback( shell, callback, clientData )
    Widget         shell;
    XtCallbackProc callback;
    XtPointer      clientData;
{
    Atom wmProtocol, wmDeleteWindow, wmSaveYourself;
    wmProtocol    =XmInternAtom( XtDisplay(shell), "WM_PROTOCOLS",     False );

    wmDeleteWindow=XmInternAtom( XtDisplay(shell), "WM_DELETE_WINDOW", False );
    wmSaveYourself=XmInternAtom( XtDisplay(shell), "WM_SAVE_YOURSELF", False );
    XmAddProtocolCallback( shell, wmProtocol, wmDeleteWindow,
			   callback, clientData );
    XmAddProtocolCallback( shell, wmProtocol, wmSaveYourself,
			   callback, clientData );
}

/*
    -- Action and callback to invoke XmpAddMwmCloseCallback
*******************************************************************************
*/
void XmpAddMwmCloseCallbackCB( w, widget_callbackList, notUsed )
    Widget	w;
    XtPointer	widget_callbackList;
    XtPointer	notUsed;
{
    char		widgetName[MAX_XRMSTRING];
    String		callbackListString;
    Widget		target;
    XtCallbackRec*	callbackList;
    XtCallbackRec*	cb;

    callbackListString = WcCleanName( (char*)widget_callbackList, widgetName );
    callbackListString = WcSkipWhitespace_Comma( callbackListString );

    target = WcFullNameToWidget( w, widgetName );

    if ( (Widget)0 == (target = WcFullNameToWidget( w, widgetName )) )
    {
	WcWARN1( w, "XmpAddMwmCloseCallback", "targetNotFound",
		"XmpAddMwmCloseCallback( %s ... ) - Target widget not found",
		widgetName );
	return;
    }

    callbackList = WcStringToCallbackList( w, callbackListString );

    for ( cb = callbackList  ;
	  cb != (XtCallbackRec*)0 && cb->callback != (XtCallbackProc)0  ;
	  cb++ )
	XmpAddMwmCloseCallback( target, cb->callback, cb->closure );

    WcFreeCallbackList( callbackList );
}

void XmpAddMwmCloseCallbackACT( w, unusedEvent, params, num_params )
    Widget      w;
    XEvent     *unusedEvent;
    String     *params;
    Cardinal   *num_params;
{
    WcInvokeCallback( XmpAddMwmCloseCallbackCB, w, params, num_params );
}

/*  -- Define a Widget as a Tab Group
*******************************************************************************
    If no widget is named, then use the widget argument.
    If the widget is a  manager widget, then this means all the children
    are in a tab group together.  If a widget is a primitive, then that
    widget is in its own tab group (this is the default).
*/

void XmpAddTabGroupACT( w, unusedEvent, params, num_params )
    Widget      w;
    XEvent     *unusedEvent;
    String     *params;
    Cardinal   *num_params;
{
    if (*num_params == 0)
    {
	XmAddTabGroup(w);
    }
    else if (*num_params == 1)
    {
	Widget target = WcFullNameToWidget( w, params[0] );

	if ((Widget)0 != target)
	{
	    XmAddTabGroup(w);
	}
	else
	{
	    WcWARN1( w, "XmpAddTabGroup", "notFound",
		"XmpAddTabGroup(%s) - widget not found", params[0] );
	}
	
    }
    else
    {
	WcWARN( w, "XmpAddTabGroup", "usage",
		"XmpAddTabGroup() requires 0 or 1 widget name argument.");
    }
}

void XmpAddTabGroupCB( w, widgetName, notUsed )
    Widget	w;
    XtPointer	widgetName;
    XtPointer	notUsed;
{
    WcInvokeAction( XmpAddTabGroupACT,  w, widgetName );
}

/*  -- XmpTable Callbacks and Actions
*******************************************************************************
    These callbacks and actions make the XmpTable convenience functions
    accessible via the Xrm database.
*/

void XmpTableChildConfigCB( w, client, ignored )
    Widget	w;
    XtPointer	client;
    XtPointer	ignored;
{
    WcInvokeAction( XmpTableChildConfigACT, w, client );
}

void XmpTableChildConfigACT( w, unusedEvent, params, num_params )
    Widget      w;
    XEvent     *unusedEvent;
    String     *params;
    Cardinal   *num_params;
{
    Widget	widget;
    Position	col;
    Position	row;
    Dimension	hSpan = 1;		/* optional arg, default */
    Dimension	vSpan = 1;		/* optional arg, default */
    XmpTableOpts opt  = TBL_DEF_OPT;	/* optional arg, default */

    if (*num_params < 3 || 6 < *num_params )
    {
	WcWARN( w, "XmpTableChildConfig", "wrongNumArgs", 
"XmpTableChildConfig( ... ) Failed - Wrong number of arguments.\n\
\tUsage: XmpTableChildConfig( w col row [h_span [v_span [opts]]] )" );
	return;
    }

    widget = WcFullNameToWidget( w, params[0] );
    col = atoi( params[1] );
    row = atoi( params[2] );

    if ((Widget)0 == widget)
    {
	WcWARN1( w, "XmpTableChildConfig", "widgetNotFound", 
"XmpTableChildConfig( %s ... ) Failed - Could not find widget.",
			params[0] );
	return;
    }
    else if ( ! XtIsSubclass( XtParent(widget), xmpTableWidgetClass ))
    {
	WcWARN1( w, "XmpTableChildConfig", "notXmpTable", 
"XmpTableChildConfig( %s ... ) Failed - Widget must be child of an XmpTable.",
			params[0] );
	return;
    }

    if ( 3 < *num_params )
	hSpan = atoi( params[3] );
    if ( 4 < *num_params )
	vSpan = atoi( params[4] );
    if ( 5 < *num_params )
	opt = XmpTableOptsParse( params[5] );

    XmpTableChildConfig( widget, col, row, hSpan, vSpan, opt );
}

void XmpTableChildPositionCB( w, client, ignored )
    Widget	w;
    XtPointer	client;
    XtPointer	ignored;
{
    WcInvokeAction( XmpTableChildPositionACT, w, client );
}

void XmpTableChildPositionACT( w, unusedEvent, params, num_params )
    Widget      w;
    XEvent     *unusedEvent;
    String     *params;
    Cardinal   *num_params;
{
    Widget	widget;
    Position	col;
    Position	row;

    if (*num_params != 3 )
    {
	WcWARN( w, "XmpTableChildPosition", "wrongNumArgs", 
"XmpTableChildPosition( ... ) Failed - Wrong number of arguments.\n\
\tUsage: XmpTableChildPosition( w col row )" );
	return;
    }

    widget = WcFullNameToWidget( w, params[0] );
    col = atoi( params[1] );
    row = atoi( params[2] );

    if ((Widget)0 == widget)
    {
	WcWARN1( w, "XmpTableChildPosition", "widgetNotFound", 
"XmpTableChildPosition( %s ... ) Failed - Could not find widget.",
			params[0] );
	return;
    }
    else if ( ! XtIsSubclass( XtParent(widget), xmpTableWidgetClass ))
    {
	WcWARN1( w, "XmpTableChildPosition", "notXmpTable", 
"XmpTableChildPosition( %s ... ) Widget is not a child of an XmpTable.",
			params[0] );
	return;
    }

    XmpTableChildPosition( widget, col, row );
}

void XmpTableChildResizeCB( w, client, ignored )
    Widget	w;
    XtPointer	client;
    XtPointer	ignored;
{
    WcInvokeAction( XmpTableChildResizeACT, w, client );
}

void XmpTableChildResizeACT( w, unusedEvent, params, num_params )
    Widget      w;
    XEvent     *unusedEvent;
    String     *params;
    Cardinal   *num_params;
{
    Widget	widget;
    Dimension	hSpan;
    Dimension	vSpan;

    if (*num_params != 3 )
    {
	WcWARN( w, "XmpTableChildResize", "wrongNumArgs", 
"XmpTableChildResize( ... ) Failed - Wrong number of arguments.\n\
\tUsage: XmpTableChildResize( w h_span v_span )" );
	return;
    }

    widget = WcFullNameToWidget( w, params[0] );
    hSpan = atoi( params[1] );
    vSpan = atoi( params[2] );

    if ((Widget)0 == widget)
    {
	WcWARN1( w, "XmpTableChildResize", "widgetNotFound", 
"XmpTableChildResize( %s ... ) Failed - Widget not found.",
			params[0] );
	return;
    }
    else if ( ! XtIsSubclass( XtParent(widget), xmpTableWidgetClass ))
    {
	WcWARN1( w, "XmpTableChildResize", "notXmpTable", 
"XmpTableChildResize( %s ... ) Failed - Widget must be child of an XmpTable.",
			params[0] );
	return;
    }

    XmpTableChildResize( widget, hSpan, vSpan );
}

void XmpTableChildOptionsCB( w, client, ignored )
    Widget	w;
    XtPointer	client;
    XtPointer	ignored;
{
    WcInvokeAction( XmpTableChildOptionsACT, w, client );
}

void XmpTableChildOptionsACT( w, unusedEvent, params, num_params )
    Widget      w;
    XEvent     *unusedEvent;
    String     *params;
    Cardinal   *num_params;
{
    Widget 	 widget;
    XmpTableOpts opt;

    if (*num_params != 2 )
    {
	WcWARN( w, "XmpTableChildOptions", "wrongNumArgs", 
"XmpTableChildOptions() Failed - Wrong number of arguments.\n\
\tUsage: XmpTableChildOptions( w options )" );
	return;
    }

    if ((Widget)0 == (widget = WcFullNameToWidget( w, params[0] )) )
    {
	WcWARN1( w, "XmpTableChildOptions", "widgetNotFound", 
"XmpTableChildOptions( %s ... ) Failed - Widget not found.",
			params[0] );
	return;
    }
    else if ( ! XtIsSubclass( XtParent(widget), xmpTableWidgetClass ))
    {
	WcWARN1( w, "XmpTableChildOptions", "notXmpTable", 
"XmpTableChildOptions( %s ... ) Failed - Widget must be child of an XmpTable.",
			params[0] );
	return;
    }

    opt = XmpTableOptsParse( params[2] );
    XmpTableChildOptions( widget, opt );
}
